tests/model/schema/schema.js

```diff
@@ -641,4 +641,87 @@ describe( 'Schema', () => {
 			expect( result[ 1 ].end.path ).to.members( [ 1 ] );
 		} );
 	} );
+
+	describe( 'getLimitElement()', () => {
+		let doc, root;
+
+		beforeEach( () => {
+			doc = new Document();
+			schema = doc.schema;
+			root = doc.createRoot();
+
+			schema.registerItem( 'div', '$block' );
+			schema.registerItem( 'article', '$block' );
+			schema.registerItem( 'section', '$block' );
+			schema.registerItem( 'paragraph', '$block' );
+			schema.allow( { name: 'paragraph', inside: 'article' } );
+			schema.allow( { name: 'article', inside: 'section' } );
+			schema.allow( { name: 'section', inside: 'div' } );
+		} );
+
+		it( 'always returns $root element if any other limit was not defined', () => {
+			schema.limits.clear();
+
+			setData( doc, '<div><section><article><paragraph>foo[]bar</paragraph></article></section></div>' );
+			expect( schema.getLimitElement( doc.selection ) ).to.equal( root );
+		} );
+
+		it( 'returns the limit element which is the closest element to common ancestor', () => {
+			schema.limits.add( 'article' );
+			schema.limits.add( 'section' );
+
+			setData( doc, '<div><section><article><paragraph>foo[]bar</paragraph></article></section></div>' );
+
+			const article = root.getNodeByPath( [ 0, 0, 0 ] );
+
+			expect( schema.getLimitElement( doc.selection ) ).to.equal( article );
+		} );
+
+		it( 'returns the limit element which is the closest element to common ancestor for non-collapsed selection', () => {
+			schema.limits.add( 'section' );
+
+			setData( doc, '<div><section><article><paragraph>fo[o</paragraph></article>b]ar</section></div>' );
+
+			const section = root.getNodeByPath( [ 0, 0 ] );
+
+			expect( schema.getLimitElement( doc.selection ) ).to.equal( section );
+		} );
+
+		it( 'works fine with multi-range selections', () => {
+			schema.limits.add( 'div' );
+
+			setData(
+				doc,
+				'<div>' +
+					'<section>' +
+						'<article>' +
+							'<paragraph>[foo]</paragraph>' +
+						'</article>' +
+					'</section>' +
+					'<section>b[]ar</section>' +
+				'</div>'
+			);
+
+			const div = root.getNodeByPath( [ 0 ] );
+			expect( schema.getLimitElement( doc.selection ) ).to.equal( div );
+		} );
+
+		it( 'works fine with multi-range selections even if limit elements are not defined', () => {
```

Comment:
If you cleared limits then this test checks nothing. The `reduce()` call may be completely broken and it will still pass. You must ensure that both ranges have some other limits and only when combined together they give some 3rd option **which still should not be the root** (so you know that it really works).